File: ios/RemoteCamPreview/Services/Network/SessionStateMachine.swift
import Foundation

enum ProtocolSessionState: String, Codable, Sendable {
    case unpaired
    case discovering
    case pairing
    case connecting
    case connected
    case interrupted
    case unavailable
    case ended
}

enum ProtocolSessionEvent: String, Codable, Sendable {
    case startDiscovery
    case peerSelected
    case pairingSucceeded
    case pairingFailed
    case transportConnected
    case transportFailed
    case controlLost
    case retry
    case capabilityLost
    case capabilityRestored
    case end
}

enum SessionStateMachineError: LocalizedError, Equatable {
    case unknownEvent(String)
    case invalidTransition(state: ProtocolSessionState, event: ProtocolSessionEvent)

    var errorDescription: String? {
        switch self {
        case .unknownEvent(let event): "UNKNOWN_STATE_EVENT: \(event)"
        case .invalidTransition(let state, let event):
            "INVALID_STATE_TRANSITION: \(event.rawValue) in \(state.rawValue)"
        }
    }
}

struct SessionStateMachine: Sendable {
    private(set) var state: ProtocolSessionState

    init(state: ProtocolSessionState = .unpaired) {
        self.state = state
    }

    mutating func apply(_ eventName: String) throws -> ProtocolSessionState {
        guard let event = ProtocolSessionEvent(rawValue: eventName) else {
            throw SessionStateMachineError.unknownEvent(eventName)
        }
        return try apply(event)
    }

    mutating func apply(_ event: ProtocolSessionEvent) throws -> ProtocolSessionState {
        let target: ProtocolSessionState?
        switch (state, event) {
        case (.unpaired, .startDiscovery): target = .discovering
        case (.discovering, .peerSelected): target = .pairing
        case (.pairing, .pairingSucceeded): target = .connecting
        case (.pairing, .pairingFailed): target = .interrupted
        case (.connecting, .transportConnected): target = .connected
        case (.connecting, .transportFailed): target = .interrupted
        case (.connected, .controlLost): target = .interrupted
        case (.interrupted, .retry): target = .discovering
        case (.unavailable, .capabilityRestored): target = .unpaired
        case (.ended, _): target = nil
        case (_, .capabilityLost): target = state == .unavailable ? nil : .unavailable
        case (_, .end): target = .ended
        default: target = nil
        }
        guard let target else {
            throw SessionStateMachineError.invalidTransition(state: state, event: event)
        }
        state = target
        return target
    }
}

File: android/app/src/main/java/dev/remotecam/preview/session/SessionStateMachine.kt
package dev.remotecam.preview.session

enum class SessionState(val wireName: String) {
    UNPAIRED("unpaired"),
    DISCOVERING("discovering"),
    PAIRING("pairing"),
    CONNECTING("connecting"),
    CONNECTED("connected"),
    INTERRUPTED("interrupted"),
    UNAVAILABLE("unavailable"),
    ENDED("ended");

    companion object {
        fun fromWireName(value: String): SessionState = entries.firstOrNull { it.wireName == value }
            ?: throw IllegalArgumentException("Unknown session state $value")
    }
}

enum class SessionEvent(val wireName: String) {
    START_DISCOVERY("startDiscovery"),
    PEER_SELECTED("peerSelected"),
    PAIRING_SUCCEEDED("pairingSucceeded"),
    PAIRING_FAILED("pairingFailed"),
    TRANSPORT_CONNECTED("transportConnected"),
    TRANSPORT_FAILED("transportFailed"),
    CONTROL_LOST("controlLost"),
    RETRY("retry"),
    CAPABILITY_LOST("capabilityLost"),
    CAPABILITY_RESTORED("capabilityRestored"),
    END("end");

    companion object {
        fun fromWireName(value: String): SessionEvent = entries.firstOrNull { it.wireName == value }
            ?: throw UnknownSessionEvent(value)
    }
}

class UnknownSessionEvent(value: String) : IllegalArgumentException("UNKNOWN_STATE_EVENT: $value")

class InvalidSessionTransition(
    val from: SessionState,
    val event: SessionEvent,
) : IllegalStateException("INVALID_STATE_TRANSITION: $from + $event")

class SessionStateMachine(initial: SessionState = SessionState.UNPAIRED) {
    var state: SessionState = initial
        private set

    fun apply(wireEvent: String): SessionState = apply(SessionEvent.fromWireName(wireEvent))

    fun apply(event: SessionEvent): SessionState {
        val target = when (state to event) {
            SessionState.UNPAIRED to SessionEvent.START_DISCOVERY -> SessionState.DISCOVERING
            SessionState.DISCOVERING to SessionEvent.PEER_SELECTED -> SessionState.PAIRING
            SessionState.PAIRING to SessionEvent.PAIRING_SUCCEEDED -> SessionState.CONNECTING
            SessionState.PAIRING to SessionEvent.PAIRING_FAILED -> SessionState.INTERRUPTED
            SessionState.CONNECTING to SessionEvent.TRANSPORT_CONNECTED -> SessionState.CONNECTED
            SessionState.CONNECTING to SessionEvent.TRANSPORT_FAILED -> SessionState.INTERRUPTED
            SessionState.CONNECTED to SessionEvent.CONTROL_LOST -> SessionState.INTERRUPTED
            SessionState.INTERRUPTED to SessionEvent.RETRY -> SessionState.DISCOVERING
            SessionState.UNAVAILABLE to SessionEvent.CAPABILITY_RESTORED -> SessionState.UNPAIRED
            else -> when {
                event == SessionEvent.CAPABILITY_LOST && state !in setOf(SessionState.UNAVAILABLE, SessionState.ENDED) ->
                    SessionState.UNAVAILABLE
                event == SessionEvent.END && state != SessionState.ENDED -> SessionState.ENDED
                else -> throw InvalidSessionTransition(state, event)
            }
        }
        state = target
        return target
    }
}
